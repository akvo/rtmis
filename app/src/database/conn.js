import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import * as SQLite from 'expo-sqlite';

const openDatabase = () => {
  const db = SQLite.openDatabase('db.db');
  return db;
};

const init = openDatabase();

const tx = (db, query, params = []) => {
  return new Promise((resolve, reject) => {
    db.transaction(
      (transaction) => {
        if (Array.isArray(query)) {
          const promises = query.map((q) => {
            return new Promise((innerResolve, innerReject) => {
              transaction.executeSql(
                q,
                params,
                (_, resultSet) => {
                  innerResolve(resultSet);
                },
                (_, error) => {
                  innerReject(error);
                  return false; // Rollback the transaction
                },
              );
            });
          });

          Promise.all(promises)
            .then((results) => {
              resolve(results);
            })
            .catch((error) => {
              reject(error);
            });
        } else {
          transaction.executeSql(
            query,
            params,
            (_, resultSet) => {
              resolve(resultSet);
            },
            (_, error) => {
              reject(error);
              return false; // Rollback the transaction
            },
          );
        }
      },
      (error) => {
        reject(error);
      },
    );
  });
};

const openDBfile = async (databaseFile, databaseName) => {
  if (!(await FileSystem.getInfoAsync(FileSystem.documentDirectory + 'SQLite')).exists) {
    await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + 'SQLite');
  }
  await FileSystem.downloadAsync(
    Asset.fromModule(databaseFile).uri,
    FileSystem.documentDirectory + `SQLite/${databaseName}.db`,
  );
  return SQLite.openDatabase(`${databaseName}.db`);
};

export const conn = {
  file: (dbFile, dbName) => openDBfile(dbFile, dbName),
  init,
  tx,
};
