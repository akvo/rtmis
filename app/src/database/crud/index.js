export { default as crudSessions } from './crud-sessions';
export { default as crudForms } from './crud-forms';
export { default as crudUsers } from './crud-users';
export { default as crudConfig } from './crud-config';
export { default as crudDataPoints } from './crud-datapoints';
