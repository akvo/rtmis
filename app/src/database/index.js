export * from './conn';
export * from './query';
export * from './tables';
