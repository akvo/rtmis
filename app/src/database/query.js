const insert = (table, data = {}) => {
  const fields = Object.keys(data);
  const valuesString = fields
    .map((key) => (isNaN(data[key]) ? `'${data[key]}'` : data[key]))
    .join(', ');
  const fieldsString = fields.join(', ');
  return `INSERT INTO ${table}(${fieldsString}) VALUES (${valuesString});`;
};

const update = (table, where = {}, data = {}) => {
  const fieldString = Object.keys(data)
    .map((key) => (isNaN(data[key]) ? `${key} = '${data[key]}'` : `${key} = ${data[key]}`))
    .join(', ');
  const conditions = Object.keys(where).map((key) => `${key} = ?`);
  const conditionString = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';
  return `UPDATE ${table} SET ${fieldString} ${conditionString};`;
};

const read = (table, where = {}, nocase = false) => {
  const conditions = Object.keys(where).map((key) => {
    return where[key] === null ? `${key} IS NULL` : `${key} = ?`;
  });
  let conditionString = '';
  if (conditions.length) {
    conditionString = `WHERE ${conditions.join(' AND ')}`;
    if (nocase) {
      conditionString += ' COLLATE NOCASE';
    }
  }
  return `SELECT * FROM ${table} ${conditionString};`;
};

const clear = (tables = []) => {
  return tables.map((t) => `DELETE FROM ${t};`);
};

const drop = (table) => {
  return `DROP TABLE IF EXISTS ${table};`;
};

const count = (table) => {
  return `SELECT COUNT(*) AS count FROM ${table}`;
};

const initialQuery = (tableName, columns) => {
  const fields = Object.keys(columns).map((key) => `${key} ${columns[key]}`);
  const fieldsString = fields.join(', ');
  return `CREATE TABLE IF NOT EXISTS ${tableName}(${fieldsString});`;
};
export const query = {
  insert,
  update,
  read,
  clear,
  drop,
  count,
  initialQuery,
};
