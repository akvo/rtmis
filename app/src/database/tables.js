export const tables = [
  {
    name: 'users',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      name: 'TEXT',
      password: 'TEXT',
      active: 'TINYINT',
    },
  },
  {
    name: 'config',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      appVersion: 'VARCHAR(255) NOT NULL',
      authenticationCode: 'TEXT',
      serverURL: 'TEXT',
      syncInterval: 'REAL',
      syncWifiOnly: 'TINYINT',
      lang: 'VARCHAR(255) DEFAULT "en" NOT NULL',
    },
  },
  {
    name: 'forms',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      formId: 'INTEGER NOT NULL',
      version: 'VARCHAR(255)',
      latest: 'TINYINT',
      name: 'VARCHAR(255)',
      json: 'TEXT',
      createdAt: 'DATETIME',
    },
  },
  {
    name: 'datapoints',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      form: 'INTEGER NOT NULL',
      user: 'INTEGER NOT NULL',
      name: 'VARCHAR(255)',
      geo: 'VARCHAR(255)',
      submitted: 'TINYINT',
      duration: 'REAL',
      createdAt: 'DATETIME',
      submittedAt: 'DATETIME',
      syncedAt: 'DATETIME',
      json: 'TEXT',
    },
  },
  {
    name: 'sessions',
    fields: {
      id: 'INTEGER PRIMARY KEY NOT NULL',
      token: 'TEXT',
      passcode: 'TEXT',
    },
  },
];
