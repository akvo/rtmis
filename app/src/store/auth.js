import { Store } from 'pullstate';
/**
 * Server config
 */
const AuthState = new Store({
  authenticationCode: '',
  useAuthenticationCode: false, // using code for authentication
  username: '',
  password: '',
  token: null,
});

export default AuthState;
