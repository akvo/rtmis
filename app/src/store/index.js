export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
