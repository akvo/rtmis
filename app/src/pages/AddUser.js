import React, { useState, useRef, useEffect } from 'react';
import { View, ToastAndroid, Platform } from 'react-native';
import { ListItem, Button, Input, Text } from '@rneui/themed';
import { Formik, ErrorMessage } from 'formik';
import * as Crypto from 'expo-crypto';
import * as Yup from 'yup';
import Icon from 'react-native-vector-icons/Ionicons';

import { BaseLayout } from '../components';
import { conn, query } from '../database';
import { UserState, UIState } from '../store';
import { i18n } from '../lib';

db = conn.init;

const AddUser = ({ navigation }) => {
  const [loading, setLoading] = useState(false);
  const [userCount, setUserCount] = useState(0);

  const formRef = useRef();
  const activeLang = UIState.useState((s) => s.lang);
  const trans = i18n.text(activeLang);
  const rightComponent = userCount ? null : false;

  const goToUsers = () => {
    navigation.navigate('Users');
  };

  const getUsersCount = async () => {
    const { rows } = await conn.tx(db, query.count('users'));
    setUserCount(rows._array?.[0]?.count);
  };

  const checkExistingUser = async (name) => {
    const checkQuery = query.read('users', { name }, true);
    const { rows } = await conn.tx(db, checkQuery, [name]);
    return rows.length;
  };

  const handleInsertDB = (data) => {
    const insertQuery = query.insert('users', data);
    conn
      .tx(db, insertQuery)
      .then(({ insertId }) => {
        if (data?.active) {
          UserState.update((s) => {
            s.id = insertId;
            s.name = data.name;
          });
        }
        if (Platform.OS === 'android') {
          ToastAndroid.show(trans.success, ToastAndroid.SHORT);
        }
        setLoading(false);
        data.active
          ? navigation.navigate('Home')
          : navigation.navigate('Users', { added: { id: insertId } });
      })
      .catch(() => {
        if (Platform.OS === 'android') {
          ToastAndroid.show(trans.errorSaveToDB, ToastAndroid.LONG);
        }
        setLoading(false);
      });
  };

  const submitData = async ({ password, name }) => {
    setLoading(true);
    const passwordEncrypted = await Crypto.digestStringAsync(
      Crypto.CryptoDigestAlgorithm.SHA1,
      password,
    );
    const isActive = userCount === 0 ? 1 : 0;
    const exist = await checkExistingUser(name);
    if (exist) {
      formRef.current.setErrors({ name: trans.errorUserExist });
      setLoading(false);
    } else {
      const data = {
        name,
        password: passwordEncrypted,
        active: isActive,
      };

      handleInsertDB(data);
      formRef.current.resetForm();
    }
  };
  const initialValues = {
    name: null,
    password: '',
    confirmPassword: null,
  };
  const addSchema = Yup.object().shape({
    name: Yup.string().required(trans.errorUserNameRequired),
    password: Yup.string().nullable(),
    confirmPassword: Yup.string().when('password', {
      is: (password) => password && password.length > 0,
      then: (schema) => schema.oneOf([Yup.ref('password')], 'Passwords must match'),
      otherwise: (schema) => schema.nullable(),
    }),
  });

  useEffect(() => {
    getUsersCount();
  }, []);

  return (
    <BaseLayout
      title={trans.addUserPageTitle}
      leftComponent={
        <Button type="clear" onPress={goToUsers} testID="arrow-back-button">
          <Icon name="arrow-back" size={18} />
        </Button>
      }
      rightComponent={rightComponent}
    >
      <Formik
        initialValues={initialValues}
        validationSchema={addSchema}
        innerRef={formRef}
        onSubmit={async (values) => {
          try {
            await submitData(values);
          } catch (err) {
            throw err;
          } finally {
            formRef.current.setSubmitting(false);
          }
        }}
      >
        {({ setFieldValue, values, handleSubmit, isSubmitting }) => (
          <BaseLayout.Content>
            <ListItem>
              <ListItem.Content>
                <ListItem.Title>
                  {`${trans.addUserInputName} `}
                  <Text color="#ff0000">*</Text>
                </ListItem.Title>
                <Input
                  placeholder={trans.addUserInputName}
                  onChangeText={(value) => setFieldValue('name', value)}
                  errorMessage={<ErrorMessage name="name" />}
                  value={values.name}
                  name="name"
                  testID="input-name"
                />
              </ListItem.Content>
            </ListItem>
            {/* <ListItem>
              <ListItem.Content>
                <ListItem.Title>Password</ListItem.Title>
                <Input
                  placeholder="Password"
                  secureTextEntry
                  onChangeText={(value) => setFieldValue('password', value)}
                  value={values.password}
                  errorMessage={<ErrorMessage name="password" />}
                  testID="input-password"
                />
              </ListItem.Content>
            </ListItem>
            <ListItem>
              <ListItem.Content>
                <ListItem.Title>Confirm Password</ListItem.Title>
                <Input
                  placeholder="Confirm Password"
                  secureTextEntry
                  onChangeText={(value) => setFieldValue('confirmPassword', value)}
                  errorMessage={<ErrorMessage name="confirmPassword" />}
                  value={values.confirmPassword}
                  testID="input-confirm-password"
                />
              </ListItem.Content>
            </ListItem> */}

            <View
              style={{ display: 'flex', flexDirection: 'column', gap: 8, paddingHorizontal: 16 }}
            >
              <Button
                onPress={handleSubmit}
                loading={loading}
                disabled={isSubmitting}
                testID="button-save"
              >
                {loading ? trans.buttonSaving : trans.buttonSave}
              </Button>
            </View>
          </BaseLayout.Content>
        )}
      </Formik>
    </BaseLayout>
  );
};

export default AddUser;
