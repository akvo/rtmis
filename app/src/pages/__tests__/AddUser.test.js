import React from 'react';
import renderer from 'react-test-renderer';
import { render, renderHook, fireEvent, act, waitFor } from '@testing-library/react-native';
import { useNavigation } from '@react-navigation/native';
import * as Crypto from 'expo-crypto';

import AddUser from '../AddUser';
import { UserState } from '../../store';
import { conn, query } from '../../database';

jest.mock('expo-sqlite');

jest.mock('expo-crypto');

db = conn.init;

describe('AddUserPage', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<AddUser />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('only username is required', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const { getByText, getByTestId } = render(<AddUser navigation={navigation} />);
    const saveButton = getByTestId('button-save');
    expect(saveButton).toBeDefined();
    fireEvent.press(saveButton);

    await waitFor(() => {
      const errorText = getByText('Username is required');
      expect(errorText).toBeDefined();
    });
  });
  /*
  test('confirm password not matched', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const { getByText, getByTestId } = render(<AddUser navigation={navigation} />);

    const passwordEl = getByTestId('input-password');
    expect(passwordEl).toBeDefined();
    const confirmPassEl = getByTestId('input-confirm-password');
    expect(confirmPassEl).toBeDefined();

    const pass1 = 'secret';
    fireEvent.changeText(passwordEl, pass1);
    const pass2 = 'Hello';
    fireEvent.changeText(confirmPassEl, pass2);

    const saveButton = getByTestId('button-save');
    expect(saveButton).toBeDefined();
    fireEvent.press(saveButton);

    await waitFor(() => {
      const errorText = getByText('Passwords must match');
      expect(errorText).toBeDefined();
    });
  }); */

  test('create username correctly', async () => {
    /**
     * Mock users count: 0
     */
    const mockCountSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: 0, _array: [{ count: 0 }] } });
    });
    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockCountSql,
      });
    });

    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const { getByTestId } = render(<AddUser navigation={navigation} />);
    const { result: userStateRef } = renderHook(() => UserState.useState((s) => s));

    const usernameEl = getByTestId('input-name');
    expect(usernameEl).toBeDefined();

    const usernameVal = 'Jhon';
    fireEvent.changeText(usernameEl, usernameVal);

    const saveButton = getByTestId('button-save');
    expect(saveButton).toBeDefined();
    fireEvent.press(saveButton);

    act(() => {
      const insertQuery = query.insert('users', { id: 1, name: usernameVal, active: 1 });
      conn.tx(db, insertQuery);

      UserState.update((s) => {
        s.id = 1;
        s.name = usernameVal;
      });
    });

    await waitFor(() => {
      const { name: usernameState } = userStateRef.current;
      expect(usernameState).toEqual(usernameVal);
      expect(navigation.navigate).toHaveBeenCalledWith('Home');
    });

    const userData = [
      {
        id: 1,
        name: usernameVal,
        active: 1,
      },
    ];
    const mockSelectSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: userData.length, _array: userData } });
    });
    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockSelectSql,
      });
    });

    const selectQuery = query.read('users', { id: 1 });
    const resultSet = await conn.tx(db, selectQuery, [1]);
    expect(resultSet.rows).toHaveLength(userData.length);
    expect(resultSet.rows._array).toEqual(userData);
  });

  it('should redirect to list users when click arrow back', () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const { getByTestId } = render(<AddUser navigation={navigation} />);

    const arrowBackEl = getByTestId('arrow-back-button');
    expect(arrowBackEl).toBeDefined();
    fireEvent.press(arrowBackEl);

    expect(navigation.navigate).toHaveBeenCalledWith('Users');
  });

  it('should be able to validate username case insensitive', async () => {
    const userData = [
      {
        id: 1,
        name: 'jHon doe',
        active: 0,
      },
    ];
    const mockSelectSql = jest.fn((query, params, successCallback) => {
      successCallback(null, { rows: { length: userData.length, _array: userData } });
    });
    db.transaction.mockImplementation((transactionFunction) => {
      transactionFunction({
        executeSql: mockSelectSql,
      });
    });

    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const { getByText, getByTestId } = render(<AddUser navigation={navigation} />);

    const name = 'Jhon Doe';
    const usernameEl = getByTestId('input-name');
    expect(usernameEl).toBeDefined();
    fireEvent.changeText(usernameEl, name);

    const saveButton = getByTestId('button-save');
    expect(saveButton).toBeDefined();
    fireEvent.press(saveButton);

    await waitFor(async () => {
      const checkQuery = query.read('users', { name }, true);
      const { rows } = await conn.tx(db, checkQuery, [name]);
      expect(rows.length).toBe(1);

      const errorText = getByText('User already exists');
      expect(errorText).toBeDefined();
    });
  });
});
