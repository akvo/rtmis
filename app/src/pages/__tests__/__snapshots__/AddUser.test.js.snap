// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserPage renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "flexDirection": "column",
      "flexWrap": "wrap",
    }
  }
  testID="stack-container"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f3f4f6",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "minheight": 78,
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="base-layout-page-title"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 2,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="arrow-back-button"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 2,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                  "paddingHorizontal": 10,
                }
              }
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontSize": 18,
              "paddingTop": 4,
              "textAlign": "center",
            }
          }
          testID="page-title"
        >
          Create New Profile
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
  <View
    style={
      {
        "backgroundColor": "#f9fafb",
        "flex": 1,
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
    testID="stack-container"
  >
    <View
      style={
        {
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#bcbbc1",
            "flexDirection": "row",
            "padding": 14,
          }
        }
        testID="RNE__LISTITEM__padView"
      >
        <View
          style={
            [
              {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
          theme={
            {
              "colors": {
                "background": "#ffffff",
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": {
                  "android": {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ad1457",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
              "spacing": {
                "lg": 12,
                "md": 8,
                "sm": 4,
                "xl": 24,
                "xs": 2,
              },
            }
          }
        >
          <Text
            accessibilityRole="text"
            style={
              {
                "backgroundColor": "transparent",
                "color": "#242424",
                "fontSize": 17,
              }
            }
            testID="listItemTitle"
          >
            Username 
            <Text
              accessibilityRole="text"
              color="#ff0000"
              style={
                {
                  "color": "#242424",
                }
              }
            >
              *
            </Text>
          </Text>
          <View
            style={
              {
                "paddingHorizontal": 10,
                "width": "100%",
              }
            }
            testID="RNE__Input__view-wrapper"
          >
            <View
              collapsable={false}
              style={
                {
                  "alignItems": "center",
                  "borderBottomWidth": 1,
                  "borderColor": "#86939e",
                  "flexDirection": "row",
                  "transform": [
                    {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <TextInput
                editable={true}
                name="name"
                onChangeText={[Function]}
                placeholder="Username"
                placeholderTextColor="#86939e"
                style={
                  {
                    "color": "#242424",
                    "flex": 1,
                    "fontSize": 18,
                    "minHeight": 40,
                  }
                }
                testID="input-name"
                underlineColorAndroid="transparent"
                value={null}
              />
            </View>
            <Text
              style={
                {
                  "color": "#ff190c",
                  "fontSize": 12,
                  "margin": 5,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "column",
          "gap": 8,
          "paddingHorizontal": 16,
          "width": "100%",
        }
      }
    >
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 2,
            },
            undefined,
            false,
          ]
        }
        testID="RNE_BUTTON_WRAPPER"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": false,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="button-save"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#2089dc",
                "borderColor": "#2089dc",
                "borderRadius": 2,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "paddingHorizontal": 10,
              }
            }
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 18,
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Save
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
